drivers/serial/uart_emul.c

```diff
@@ -0,0 +1,181 @@
+/*
+ * Copyright (c) 2023 Fabian Blatz
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#define DT_DRV_COMPAT zephyr_uart_emul
+
+#include <zephyr/kernel.h>
+#include <zephyr/drivers/uart.h>
+#include <zephyr/sys/ring_buffer.h>
+#include <zephyr/drivers/serial/uart_emul.h>
+#include <errno.h>
+
+#include <zephyr/logging/log.h>
+LOG_MODULE_REGISTER(uart_emul, CONFIG_UART_LOG_LEVEL);
+
+struct uart_emul_config {
+	bool loopback;
+};
+
+/* Device run time data */
+struct uart_emul_data {
+	struct uart_config cfg;
+
+	struct ring_buf *rx_rb;
+	struct k_spinlock rx_lock;
+
+	uart_emul_callback_tx_data_ready_t tx_data_ready_cb;
+	struct ring_buf *tx_rb;
+	struct k_spinlock tx_lock;
+};
+
+int uart_emul_poll_in(const struct device *dev, unsigned char *p_char)
+{
+	struct uart_emul_data *drv_data = dev->data;
+	k_spinlock_key_t key;
+
+	key = k_spin_lock(&drv_data->rx_lock);
+	if (ring_buf_is_empty(drv_data->rx_rb)) {
+		LOG_DBG("Rx buffer is empty");
+		k_spin_unlock(&drv_data->rx_lock, key);
+		return -ENOBUFS;
+	}
+	ring_buf_get(drv_data->rx_rb, p_char, 1);
+	k_spin_unlock(&drv_data->rx_lock, key);
+	return 0;
+}
+
+void uart_emul_poll_out(const struct device *dev, unsigned char out_char)
+{
+	struct uart_emul_data *drv_data = dev->data;
+	const struct uart_emul_config *drv_cfg = dev->config;
+	k_spinlock_key_t key;
+
+	key = k_spin_lock(&drv_data->tx_lock);
+	if (!ring_buf_put(drv_data->tx_rb, &out_char, 1)) {
+		LOG_DBG("Tx buffer is full");
+	}
+	k_spin_unlock(&drv_data->tx_lock, key);
+	if (drv_cfg->loopback) {
+		uart_emul_put_rx_data(dev, &out_char, 1);
+	}
+	if (drv_data->tx_data_ready_cb) {
+		(drv_data->tx_data_ready_cb)(dev, ring_buf_size_get(drv_data->tx_rb));
```

Comment:
We can't call the ready callback if the `ring_buf_put` failed, right?

Reply:
The buffer is still full, so I did not do an early return. This way the application/test is prompted again to take care of the tx buffer contents. It's a bit unfortunate that the UART API for `poll_out` does not have a return value.

Reply:
Thought about it - if the application/tests previously failed to take care of the buffer, it is unlikely that subsequent callback calls yield different behaviour.